accepted/2024/trusted-publishers-oidc-for-nuget-push.technical.md

```diff
@@ -0,0 +1,189 @@
+  - This is to avoid resurrection attacks.
+- If a branch filter is provided in the trust policy:
+  - `ref_type` claim must be `branch`
+  - `ref` claim must be `refs/head/{branch}` (case insensitive)
+- If an environment filter is provided in the trust policy:
+  - `environment` claim must be `{environment}` (case insensitive)
+- If a workflow path filter is provided in the trust policy:
+   - `job_workflow_ref` claim must be `{repo owner}/{repo name}/{workflow path}@.*` (case insensitive)
+   - The workflow path should be normalized to `/` path separators at the time of trust policy creation
+
+A list of possible claims to verify against is available in GitHub's [Understanding the OIDC
+token](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token)
+document.
+
+Other Trusted Publishers like Azure DevOps or Bitbucket should have sufficient token claims so both the registry
+(NuGet.org) and the package author are certain that only proper workload identity tokens are traded for privileged
+short-lived API keys.
+
+## Data relationships (persisting trust policies, schema changes)
+
+A new SQL table will be added to the NuGetGallery database to store trust policies. The table should be generic enough
+to allow us to add additional Trusted Publishers without a DB schema change (ideally). The trust policy DB record with
+have a foreign key to the `Users` table (containing both user and organization records) but will be restricted from
+associating with organization records by the application. Many of the columns will be shared with the `Credentials`
+table in order to express the scoping rules to be copied into the short-lived API key.
+
+A nullable column will be added to the `Credentials` table to refer to the trust policy used to create the short-lived
+API key. This will allow package publish operations to audit their related trust policy information.
+
+Deleting a trust policy should have the effect of deleting all related short-lived API keys.
+
+## Trading an OIDC token for an API key
+
+A new endpoint will be needed for trading an bearer token (OIDC token, a JWT) for an API key. The endpoint URL will be
+discoverable via the [V3 service index](https://learn.microsoft.com/en-us/nuget/api/overview#service-index) and a new
+resource type which is `ApiKeyService/1.0.0`. For NuGet.org, the service index is available at
+`https://api.nuget.org/v3/index.json`. The new resource URL be something like `https://www.nuget.org/api/v2/api-key`.
+
+```
+POST /api/v2/api-key HTTP/1.1
+Host: www.nuget.org
+Authorization: Bearer {OIDC token}
+Content-Type: application/json
+
+{
+   "username": "{username of user with trust policy}"
+}
+```
+
+The response will look like this:
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/json
+
+{
+   "api_key": "{short lived API key in clear text}",
```

Comment:
I didn't see any explicit mention of how long the short-lived API key will be valid.  I think that's intentional, since we should have some flexibility in adjusting its lifespan.  However, for the sake of argument, what is a reasonable starting point?  30 minutes?

What is the user experience if the short-lived API key expires before they're done publishing and how do they recover?  Or, is the client (NuGet client, here) supposed to automatically reacquire a new API key transparently so that the user never notices the expiration?

I'm trying to understand where this automagical flow might actually break down for users.

Reply:
> I didn't see any explicit mention of how long the short-lived API key will be valid. I think that's intentional, since we should have some flexibility in adjusting its lifespan. However, for the sake of argument, what is a reasonable starting point? 30 minutes?

I actually thought I documented. I did not 😂. I was planning on using 15 minutes. This is what PyPI uses and it should have good coverage for our push patterns. I will add supporting data.